libfs/file.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2017 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libfs
+
+import (
+	"io"
+	"sync/atomic"
+
+	"github.com/keybase/kbfs/libkbfs"
+	"github.com/pkg/errors"
+	billy "github.com/src-d/go-billy"
+)
+
+// File is a wrapper around a libkbfs.Node that implements the
+// billy.File interface.
+type File struct {
+	fs       *FS
+	filename string
+	node     libkbfs.Node
+	readOnly bool
+	offset   int64
+}
+
+var _ billy.File = (*File)(nil)
+
+// Name implements the billy.File interface for File.
+func (f *File) Name() string {
+	return f.filename
+}
+
+func (f *File) updateOffset(origOffset, advanceBytes int64) {
+	// If there are two concurrent Write calls at the same time, it's
+	// not well-defined what the offset should be after.  Just set it
+	// to what this call thinks it should be and let the application
+	// sort things out.
+	_ = atomic.SwapInt64(&f.offset, origOffset+advanceBytes)
+}
+
+// Write implements the billy.File interface for File.
+func (f *File) Write(p []byte) (n int, err error) {
+	if f.readOnly {
+		return 0, errors.New("Trying to write a read-only file")
+	}
+
+	origOffset := atomic.LoadInt64(&f.offset)
```

Comment:
I didn't want to hold the mutex over the blocking Read/Write operations, and I was finding it annoying to read the value under the lock early in the function, because I couldn't use `defer` to release the lock.  This way seemed like less code, and was still clear.  But if you don't like it, I could change it to mutex.